spec/publisher_spec.rb

```diff
@@ -71,5 +71,69 @@
         Capistrano::S3::Publisher.publish!('s3.amazonaws.com', 'abc', '123', 'mybucket.amazonaws.com', 'spec/sample', '', 'cf123', [], exclude_paths, false, {}, 'staging')
       end
     end
+
+    context "write options" do
+      it "sets bucket write options to all files" do
+        headers = { cache_control: 'no-cache' }
+        extra_options = { write: headers }
+
+        Aws::S3::Client.any_instance.expects(:put_object).with() { |options| contains(options, headers) }.times(3)
+        Capistrano::S3::Publisher.publish!('s3.amazonaws.com', 'abc', '123', 'mybucket.amazonaws.com', 'spec/sample-write', '', 'cf123', [], [], false, extra_options, 'staging')
+      end
+
+      it "sets object write options to a single file" do
+        headers = { cache_control: 'no-cache', acl: :private }
+        extra_options = {
+          object_write: {
+            'index.html' => headers
+          }
+        }
+
+        Aws::S3::Client.any_instance.expects(:put_object).with { |options| options[:key] == 'index.html' && contains(options, headers) }.once
+        Aws::S3::Client.any_instance.expects(:put_object).with { |options| options[:key] != 'index.html' && !contains(options, headers) }.twice
```

Comment:
This is legacy syntax, we should prefer the new one that doesn't monkey patch all classes; 

```ruby
expect_any_instance_of(Aws::S3::Client).to receive(:put_object).with...
```

https://relishapp.com/rspec/rspec-mocks/docs/old-syntax/any-instance

Reply:
Agreed, but let’s do it separately from this PR to keep the code consistent.

Reply:
Legit!